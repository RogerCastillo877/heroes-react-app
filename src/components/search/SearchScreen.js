import { useLocation, useNavigate } from 'react-router-dom';
import { useMemo } from 'react';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHerosByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const { q = ''} = queryString.parse( location.search );

  const [ formValues, handleInputChange ] = useForm({
    searchText: q
  });

  const { searchText } = formValues;
  const heroesFilted = useMemo( () => getHeroesByName( q ), [ q ] );

  const handleSearch = (e) => {
    e.preventDefault();

    navigate(`?q=${ searchText }`);
  };

  return (
    <>
      <h1>SearchScreen</h1>
      <hr />
      <div className='row'>
        <div className='col-5'>
          <h4>Buscar</h4>
          <hr />

          <form onSubmit={ handleSearch } aria-label= "form">
            <input
              type='text'
              placeholder='Buscar un héroe'
              className='form-control'
              name='searchText'
              autoComplete='off'
              value={ searchText }
              onChange={ handleInputChange }
            />

            <button
              className='btn btn-outline-primary mt-1'
              type='submit'
            >
              Buscar
            </button>
          </form>
        </div>

        <div className='col-7'>
          <h4>Resultados</h4>
          <hr />
          {
            ( q === '' )
            ? <div className='alert alert-info'>Buscar un héroe</div>
            : ( heroesFilted.length === 0 ) && <div className='alert alert-danger' aria-label='alert-danger'>No hay resultados: { q }</div>
          }
          {
            heroesFilted.map( hero => (
              <HeroCard
                key={ hero.id }
                { ...hero }
              />
            ))
          }
        </div>
      </div>
    </>
  )
}
