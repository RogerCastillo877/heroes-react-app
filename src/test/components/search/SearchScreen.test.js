import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('should test searchScreen.js', () => {

  beforeEach( () => jest.clearAllMocks() );

  test('should display properly with default values', () => {

    const { container } = render(
      <MemoryRouter>
        <SearchScreen />
      </MemoryRouter>
    )

    expect( container ).toMatchSnapshot();
  });

  test('should display to batman and input value of querySting', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox');
    expect( input.value ).toBe('batman');

    const img = screen.getByRole('img');
    expect( img.src ).toContain('/assets/dc-batman.jpg')

    // const alert = screen.getByLabelText('alert-danger');
    // expect( alert.style.display ).toBe('none');
  });

  test('should display an error if don\'t find a hero (batman123)', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    )

    const alert = screen.getByLabelText('alert-danger');
    expect( alert.style.display ).toBe('');
  });

  test('should call navigate to the new screen', () => {

    const inputValue = 'superman';

    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change( input, { target: { name: 'searchText', value: inputValue } });
    
    const form = screen.getByLabelText('form');
    fireEvent.submit( form );

    expect( mockedUseNavigate ).toHaveBeenCalledWith(`?q=${ inputValue }`);
  });
});
