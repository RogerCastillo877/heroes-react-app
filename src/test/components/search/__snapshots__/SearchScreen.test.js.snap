// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test searchScreen.js should display properly with default values 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>
</div>
`;
